sale_loyalty_order_line_link/models/sale_order.py

```diff
@@ -1,41 +1,22 @@
 # Copyright 2021 Tecnativa - David Vidal
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 from odoo import fields, models
-from odoo.tools.safe_eval import safe_eval
 
 
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
-    def _get_reward_values_product(self, program):
-        """Add the link to the program in the discount line"""
-        res = super()._get_reward_values_product(program)
-        res["coupon_program_id"] = program.id
-        return res
-
-    def _get_reward_values_discount(self, program):
-        """Add the link to the program in the discount line. The original method returns
-        a dict.values() to support multiple taxes. We can safely return a list instead
-        as the main method just iterates over the result of these auxiliar products to
-        write vals. https://git.io/J88As
-        """
-        res = list(super()._get_reward_values_discount(program))
-        [r.update(coupon_program_id=program.id) for r in res]
-        return res
-
     def write(self, vals):
         """We're looking for reward lines set in the write, from those, we'll get
         the proper ids to write them into the related lines"""
         res = super().write(vals)
-        reward_lines = [
-            x[2]
-            for x in vals.get("order_line", [])
-            if len(x) > 2 and x[0] == 0 and x[2].get("is_reward_line")
-        ]
+        reward_lines = self.order_line.filtered(
+            lambda x: x.loyalty_program_id and x.is_reward_line
+        )
         if not reward_lines:
             return res
-        programs = self.env["coupon.program"].browse(
-            list({x.get("coupon_program_id") for x in reward_lines})
+        programs = self.env["loyalty.program"].browse(
+            list({x.loyalty_program_id.id for x in reward_lines})
```

Comment:
Don't rebrowse here, but just populate the recordset:

```suggestion
        programs = reward_lines.loyalty_program_id
```